File: facebook-common/src/main/java/com/facebook/login/CustomTabLoginMethodHandler.java
/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook.login;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ResolveInfo;
import android.content.pm.ServiceInfo;
import android.net.Uri;
import android.os.Bundle;
import android.os.Parcel;
import android.os.Parcelable;

import com.facebook.AccessTokenSource;
import com.facebook.CustomTabMainActivity;
import com.facebook.FacebookException;
import com.facebook.FacebookOperationCanceledException;
import com.facebook.FacebookRequestError;
import com.facebook.FacebookSdk;
import com.facebook.FacebookServiceException;
import com.facebook.internal.FetchedAppSettings;
import com.facebook.internal.FetchedAppSettingsManager;
import com.facebook.internal.ServerProtocol;
import com.facebook.internal.Utility;
import com.facebook.internal.Validate;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class CustomTabLoginMethodHandler extends WebLoginMethodHandler {
    private static final int CUSTOM_TAB_REQUEST_CODE = 1;
    private static final int CHALLENGE_LENGTH = 20;
    private static final int API_EC_DIALOG_CANCEL = 4201;
    private static final String CUSTOM_TABS_SERVICE_ACTION =
            "android.support.customtabs.action.CustomTabsService";
    private static final String[] CHROME_PACKAGES = {
            "com.android.chrome",
            "com.chrome.beta",
            "com.chrome.dev",
    };

    private String currentPackage;
    private String expectedChallenge;

    CustomTabLoginMethodHandler(LoginClient loginClient) {
        super(loginClient);
        expectedChallenge = Utility.generateRandomString(CHALLENGE_LENGTH);
    }

    @Override
    String getNameForLogging() {
        return "custom_tab";
    }

    @Override
    AccessTokenSource getTokenSource() {
        return AccessTokenSource.CHROME_CUSTOM_TAB;
    }

    @Override
    protected String getSSODevice() {
        return "chrome_custom_tab";
    }

    @Override
    boolean tryAuthorize(final LoginClient.Request request) {
        if (!isCustomTabsAllowed()) {
            return false;
        }

        Bundle parameters = getParameters(request);
        parameters = addExtraParameters(parameters, request);
        Activity activity = loginClient.getActivity();

        Intent intent = new Intent(activity, CustomTabMainActivity.class);
        intent.putExtra(CustomTabMainActivity.EXTRA_PARAMS, parameters);
        intent.putExtra(CustomTabMainActivity.EXTRA_CHROME_PACKAGE, getChromePackage());
        loginClient.getFragment().startActivityForResult(intent, CUSTOM_TAB_REQUEST_CODE);

        return true;
    }

    private boolean isCustomTabsAllowed() {
        return isCustomTabsEnabled()
                && getChromePackage() != null
                && Validate.hasCustomTabRedirectActivity(FacebookSdk.getApplicationContext());
    }

    private boolean isCustomTabsEnabled() {
        final String appId = Utility.getMetadataApplicationId(loginClient.getActivity());
        final FetchedAppSettings settings = FetchedAppSettingsManager.getAppSettingsWithoutQuery(appId);
        return settings != null && settings.getCustomTabsEnabled();
    }

    private String getChromePackage() {
        if (currentPackage != null) {
            return currentPackage;
        }
        Context context = loginClient.getActivity();
        Intent serviceIntent = new Intent(CUSTOM_TABS_SERVICE_ACTION);
        List<ResolveInfo> resolveInfos =
                context.getPackageManager().queryIntentServices(serviceIntent, 0);
        if (resolveInfos != null) {
            Set<String> chromePackages = new HashSet<>(Arrays.asList(CHROME_PACKAGES));
            for (ResolveInfo resolveInfo : resolveInfos) {
                ServiceInfo serviceInfo = resolveInfo.serviceInfo;
                if (serviceInfo != null && chromePackages.contains(serviceInfo.packageName)) {
                    currentPackage = serviceInfo.packageName;
                    return currentPackage;
                }
            }
        }
        return null;
    }

    @Override
    boolean onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode != CUSTOM_TAB_REQUEST_CODE) {
            return super.onActivityResult(requestCode, resultCode, data);
        }
        LoginClient.Request request = loginClient.getPendingRequest();
        if (resultCode == Activity.RESULT_OK) {
            onCustomTabComplete(data.getStringExtra(CustomTabMainActivity.EXTRA_URL), request);
            return true;
        }
        super.onComplete(request, null, new FacebookOperationCanceledException());
        return false;
    }

    private void onCustomTabComplete(String url, LoginClient.Request request) {
        if (url != null && url.startsWith(CustomTabMainActivity.getRedirectUrl())) {
            Uri uri = Uri.parse(url);
            Bundle values = Utility.parseUrlQueryString(uri.getQuery());
            values.putAll(Utility.parseUrlQueryString(uri.getFragment()));

            if (!validateChallengeParam(values)) {
                super.onComplete(request, null, new FacebookException("Invalid state parameter"));
                return;
            }

            String error = values.getString("error");
            if (error == null) {
                error = values.getString("error_type");
            }

            String errorMessage = values.getString("error_msg");
            if (errorMessage == null) {
                errorMessage = values.getString("error_message");
            }
            if (errorMessage == null) {
                errorMessage = values.getString("error_description");
            }
            String errorCodeString = values.getString("error_code");
            int errorCode = FacebookRequestError.INVALID_ERROR_CODE;
            if (!Utility.isNullOrEmpty(errorCodeString)) {
                try {
                    errorCode = Integer.parseInt(errorCodeString);
                } catch (NumberFormatException ex) {
                    errorCode = FacebookRequestError.INVALID_ERROR_CODE;
                }
            }

            if (Utility.isNullOrEmpty(error) && Utility.isNullOrEmpty(errorMessage)
                    && errorCode == FacebookRequestError.INVALID_ERROR_CODE) {
                super.onComplete(request, values, null);
            } else if (error != null && (error.equals("access_denied") ||
                    error.equals("OAuthAccessDeniedException"))) {
                super.onComplete(request, null, new FacebookOperationCanceledException());
            } else if (errorCode == API_EC_DIALOG_CANCEL) {
                super.onComplete(request, null, new FacebookOperationCanceledException());
            } else {
                FacebookRequestError requestError =
                        new FacebookRequestError(errorCode, error, errorMessage);
                super.onComplete(
                        request,
                        null,
                        new FacebookServiceException(requestError, errorMessage));
            }
        }
    }

    @Override
    protected void putChallengeParam(JSONObject param) throws JSONException {
        param.put(LoginLogger.EVENT_PARAM_CHALLENGE, expectedChallenge);
    }

    private boolean validateChallengeParam(Bundle values) {
        try {
            String stateString = values.getString(ServerProtocol.DIALOG_PARAM_STATE);
            if (stateString == null) {
                return false;
            }
            JSONObject state = new JSONObject(stateString);
            String challenge = state.getString(LoginLogger.EVENT_PARAM_CHALLENGE);
            return challenge.equals(expectedChallenge);
        } catch (JSONException e) {
            return false;
        }
    }

    @Override
    public int describeContents() {
        return 0;
    }

    CustomTabLoginMethodHandler(Parcel source) {
        super(source);
        expectedChallenge = source.readString();
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        dest.writeString(expectedChallenge);
    }

    public static final Parcelable.Creator<CustomTabLoginMethodHandler> CREATOR =
            new Parcelable.Creator() {

                @Override
                public CustomTabLoginMethodHandler createFromParcel(Parcel source) {
                    return new CustomTabLoginMethodHandler(source);
                }

                @Override
                public CustomTabLoginMethodHandler[] newArray(int size) {
                    return new CustomTabLoginMethodHandler[size];
                }
            };
}

File: facebook-common/src/main/java/com/facebook/login/CustomTabLoginMethodHandler.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.login

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Parcel
import android.os.Parcelable
import com.facebook.AccessTokenSource
import com.facebook.CustomTabMainActivity
import com.facebook.FacebookException
import com.facebook.FacebookOperationCanceledException
import com.facebook.FacebookRequestError
import com.facebook.FacebookSdk
import com.facebook.FacebookSdk.getExecutor
import com.facebook.FacebookServiceException
import com.facebook.internal.CustomTab.Companion.getURIForAction
import com.facebook.internal.CustomTabUtils.getChromePackage
import com.facebook.internal.CustomTabUtils.getValidRedirectURI
import com.facebook.internal.InstagramCustomTab
import com.facebook.internal.ServerProtocol
import com.facebook.internal.Utility.generateRandomString
import com.facebook.internal.Utility.isNullOrEmpty
import com.facebook.internal.Utility.parseUrlQueryString
import com.facebook.internal.Validate
import com.facebook.login.LoginClient.Request
import java.lang.NumberFormatException
import org.json.JSONException
import org.json.JSONObject

class CustomTabLoginMethodHandler : WebLoginMethodHandler {
  private var currentPackage: String? = null
  private var expectedChallenge: String?
  private var validRedirectURI: String

  constructor(loginClient: LoginClient) : super(loginClient) {
    expectedChallenge = generateRandomString(CHALLENGE_LENGTH)
    calledThroughLoggedOutAppSwitch = false
    validRedirectURI = getValidRedirectURI(developerDefinedRedirectURI)
  }

  override val nameForLogging: String = "custom_tab"
  override val tokenSource: AccessTokenSource = AccessTokenSource.CHROME_CUSTOM_TAB
  private val developerDefinedRedirectURI: String
    get() = super.getRedirectUrl()

  override fun getRedirectUrl(): String = validRedirectURI

  override fun getSSODevice(): String? = "chrome_custom_tab"

  override fun tryAuthorize(request: Request): Int {
    val loginClient = loginClient
    if (getRedirectUrl().isEmpty()) {
      return 0
    }
    var parameters = getParameters(request)
    parameters = addExtraParameters(parameters, request)
    if (calledThroughLoggedOutAppSwitch) {
      parameters.putString(ServerProtocol.DIALOG_PARAM_CCT_OVER_LOGGED_OUT_APP_SWITCH, "1")
    }
    if (FacebookSdk.hasCustomTabsPrefetching) {
      if (request.isInstagramLogin) {
        CustomTabPrefetchHelper.mayLaunchUrl(
            InstagramCustomTab.getURIForAction(OAUTH_DIALOG, parameters))
      } else {
        CustomTabPrefetchHelper.mayLaunchUrl(getURIForAction(OAUTH_DIALOG, parameters))
      }
    }
    val activity: Activity = loginClient.activity ?: return 0
    val intent = Intent(activity, CustomTabMainActivity::class.java)
    intent.putExtra(CustomTabMainActivity.EXTRA_ACTION, OAUTH_DIALOG)
    intent.putExtra(CustomTabMainActivity.EXTRA_PARAMS, parameters)
    intent.putExtra(CustomTabMainActivity.EXTRA_CHROME_PACKAGE, chromePackage)
    intent.putExtra(CustomTabMainActivity.EXTRA_TARGET_APP, request.loginTargetApp.toString())
    loginClient.fragment?.startActivityForResult(intent, CUSTOM_TAB_REQUEST_CODE)
    return 1
  }

  private val chromePackage: String?
    get() {
      if (currentPackage != null) {
        return currentPackage
      }
      currentPackage = getChromePackage()
      return currentPackage
    }

  override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
    if (data != null) {
      val hasNoBrowserException =
          data.getBooleanExtra(CustomTabMainActivity.NO_ACTIVITY_EXCEPTION, false)
      if (hasNoBrowserException) {
        return super.onActivityResult(requestCode, resultCode, data)
      }
    }
    if (requestCode != CUSTOM_TAB_REQUEST_CODE) {
      return super.onActivityResult(requestCode, resultCode, data)
    }
    val request = loginClient.pendingRequest ?: return false
    if (resultCode == Activity.RESULT_OK) {
      var extraUrl: String? = null
      if (data != null) {
        extraUrl = data.getStringExtra(CustomTabMainActivity.EXTRA_URL)
      }
      onCustomTabComplete(extraUrl, request)
      return true
    }
    super.onComplete(request, null, FacebookOperationCanceledException())
    return false
  }

  private fun onCustomTabComplete(url: String?, request: LoginClient.Request) {
    if (url != null &&
        (url.startsWith(Validate.CUSTOM_TAB_REDIRECT_URI_PREFIX) ||
            url.startsWith(super.getRedirectUrl()))) {
      val uri = Uri.parse(url)
      val values = parseUrlQueryString(uri.query)
      values.putAll(parseUrlQueryString(uri.fragment))
      if (!validateChallengeParam(values)) {
        super.onComplete(request, null, FacebookException("Invalid state parameter"))
        return
      }
      var error = values.getString("error")
      if (error == null) {
        error = values.getString("error_type")
      }
      var errorMessage = values.getString("error_msg")
      if (errorMessage == null) {
        errorMessage = values.getString("error_message")
      }
      if (errorMessage == null) {
        errorMessage = values.getString("error_description")
      }
      val errorCodeString = values.getString("error_code")
      val errorCode =
          try {
            errorCodeString?.toInt() ?: FacebookRequestError.INVALID_ERROR_CODE
          } catch (ex: NumberFormatException) {
            FacebookRequestError.INVALID_ERROR_CODE
          }
      if (isNullOrEmpty(error) &&
          isNullOrEmpty(errorMessage) &&
          errorCode == FacebookRequestError.INVALID_ERROR_CODE) {
        if (values.containsKey("access_token")) {
          super.onComplete(request, values, null)
          return
        }
        getExecutor().execute {
          try {
            val processedValues = processCodeExchange(request, values)
            onComplete(request, processedValues, null)
          } catch (ex: FacebookException) {
            onComplete(request, null, ex)
          }
        }
      } else if (error != null &&
          (error == "access_denied" || error == "OAuthAccessDeniedException")) {
        super.onComplete(request, null, FacebookOperationCanceledException())
      } else if (errorCode == API_EC_DIALOG_CANCEL) {
        super.onComplete(request, null, FacebookOperationCanceledException())
      } else {
        val requestError = FacebookRequestError(errorCode, error, errorMessage)
        super.onComplete(request, null, FacebookServiceException(requestError, errorMessage))
      }
    }
  }

  @Throws(JSONException::class)
  override fun putChallengeParam(param: JSONObject) {
    param.put(LoginLogger.EVENT_PARAM_CHALLENGE, expectedChallenge)
  }

  private fun validateChallengeParam(values: Bundle): Boolean {
    return try {
      val stateString = values.getString(ServerProtocol.DIALOG_PARAM_STATE) ?: return false
      val state = JSONObject(stateString)
      val challenge = state.getString(LoginLogger.EVENT_PARAM_CHALLENGE)
      challenge == expectedChallenge
    } catch (e: JSONException) {
      false
    }
  }

  override fun describeContents(): Int = 0

  internal constructor(source: Parcel) : super(source) {
    expectedChallenge = source.readString()
    validRedirectURI = getValidRedirectURI(developerDefinedRedirectURI)
  }

  override fun writeToParcel(dest: Parcel, flags: Int) {
    super.writeToParcel(dest, flags)
    dest.writeString(expectedChallenge)
  }

  companion object {
    private const val CUSTOM_TAB_REQUEST_CODE = 1
    private const val CHALLENGE_LENGTH = 20
    private const val API_EC_DIALOG_CANCEL = 4201
    const val OAUTH_DIALOG = "oauth"
    @JvmField var calledThroughLoggedOutAppSwitch = false
    @JvmField
    val CREATOR: Parcelable.Creator<CustomTabLoginMethodHandler> =
        object : Parcelable.Creator<CustomTabLoginMethodHandler> {
          override fun createFromParcel(source: Parcel): CustomTabLoginMethodHandler {
            return CustomTabLoginMethodHandler(source)
          }

          override fun newArray(size: Int): Array<CustomTabLoginMethodHandler?> {
            return arrayOfNulls(size)
          }
        }
  }
}
